Serialized optionals and instants cause issues in modern Java

OCC will require a modern version of Java soon to continue supporting modern versions of Minecraft. Community does not fully support at the moment. We can work around the below error, however it would be best to be fixed in Community if at all possible. This below error happens when broadcasting punishments globally.

```
[03:38:39 INFO]: mameBT issued server command: /ban amdgpu test
[03:38:41 INFO]: [A] *❖❖mameBT » Permanent Ban » AMDGPU » test

[03:38:41 WARN]: Exception in thread "Craft Scheduler Thread - 11"
[03:38:41 WARN]: org.apache.commons.lang.UnhandledException: Plugin Community v0.2-SNAPSHOT (git-a7b0506) generated an exception while executing task 238
        at org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57)
        at org.github.paperspigot.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:23)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.google.gson.JsonIOException: Failed making field 'java.util.Optional#value' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
        at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:38)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:286)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
        at com.google.gson.Gson.getAdapter(Gson.java:556)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.createBoundField(ReflectiveTypeAdapterFactory.java:160)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:294)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
        at com.google.gson.Gson.getAdapter(Gson.java:556)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.createBoundField(ReflectiveTypeAdapterFactory.java:160)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:294)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
        at com.google.gson.Gson.getAdapter(Gson.java:556)
        at com.google.gson.Gson.toJson(Gson.java:834)
        at com.google.gson.Gson.toJson(Gson.java:812)
        at com.google.gson.Gson.toJson(Gson.java:759)
        at com.google.gson.Gson.toJson(Gson.java:736)
        at dev.pgm.community.network.updates.NetworkUpdateBase.getData(NetworkUpdateBase.java:24)
        at dev.pgm.community.network.types.RedisNetworkFeature.lambda$sendUpdate$1(RedisNetworkFeature.java:100)
        at org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftTask.run(CraftTask.java:64)
        at org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:53)
        ... 4 more
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.lang.Object java.util.Optional.value accessible: module java.base does not "opens java.util" to unnamed module @76b10754
        at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
        at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:183)
        at java.base/java.lang.reflect.Field.setAccessible(Field.java:177)
        at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:35)
        ... 23 more
```

--- a/src/main/java/dev/pgm/community/moderation/commands/MuteCommand.java
+++ b/src/main/java/dev/pgm/community/moderation/commands/MuteCommand.java
@@ -86,7 +86,7 @@ public class MuteCommand extends CommunityCommand {
                                   name -> {
                                     if (isMuted.isPresent()) {
                                       moderation
-                                          .unmute(id.get(), audience.getId())
+                                          .unmute(id.get(), audience.getId().orElse(null))
                                           .thenAcceptAsync(
                                               pardon -> {
                                                 if (!pardon) {

--- a/src/main/java/dev/pgm/community/moderation/commands/PunishmentCommand.java
+++ b/src/main/java/dev/pgm/community/moderation/commands/PunishmentCommand.java
@@ -127,7 +127,7 @@ public class PunishmentCommand extends CommunityCommand {
             isBanned -> {
               if (isBanned) {
                 moderation
-                    .pardon(target.getIdentifier(), audience.getId())
+                    .pardon(target.getIdentifier(), audience.getId().orElse(null))
                     .thenAcceptAsync(
                         pardon -> {
                           if (!pardon) {

--- a/src/main/java/dev/pgm/community/moderation/feature/ModerationFeature.java
+++ b/src/main/java/dev/pgm/community/moderation/feature/ModerationFeature.java
@@ -55,10 +55,10 @@ public interface ModerationFeature extends Feature {
    * Pardon target for past punishments (Ban/Tempban)
    *
    * @param target A username or UUID string
-   * @param issuer An optional UUID of the command sender (console is empty)
+   * @param issuer A UUID of the command sender (null for console)
    * @return True if any ban infractions were lifted, false if none
    */
-  CompletableFuture<Boolean> pardon(String target, Optional<UUID> issuer);
+  CompletableFuture<Boolean> pardon(String target, @Nullable UUID issuer);
 
   /**
    * Deactivate an active punishment
@@ -97,10 +97,10 @@ public interface ModerationFeature extends Feature {
    * Unmutes any active mutes for the provided target
    *
    * @param target A player UUID
-   * @param issuer The person lifting the infraction
+   * @param issuer The UUID of the player lifting the mute (null for console)
    * @return true if unmute was removed, false if no mute existed
    */
-  CompletableFuture<Boolean> unmute(UUID target, Optional<UUID> issuer);
+  CompletableFuture<Boolean> unmute(UUID target, @Nullable UUID issuer);
 
   /**
    * Gets a set of online players who are muted

--- a/src/main/java/dev/pgm/community/moderation/feature/ModerationFeatureBase.java
+++ b/src/main/java/dev/pgm/community/moderation/feature/ModerationFeatureBase.java
@@ -138,11 +138,11 @@ public abstract class ModerationFeatureBase extends FeatureBase implements Moder
             target,
             getSenderId(issuer.getSender()),
             reason,
-            time,
+            time.toEpochMilli(),
             duration,
             type,
             active,
-            time,
+            time.toEpochMilli(),
             getSenderId(issuer.getSender()),
             getModerationConfig().getService());
     Bukkit.getPluginManager().callEvent(new PlayerPunishmentEvent(issuer, punishment, silent));
@@ -157,7 +157,7 @@ public abstract class ModerationFeatureBase extends FeatureBase implements Moder
   @Override
   public Optional<Punishment> getLastPunishment(UUID issuer) {
     return recents.stream()
-        .filter(p -> p.getIssuerId().isPresent() && p.getIssuerId().get().equals(issuer))
+        .filter(p -> !p.isConsole() && p.getIssuerId().equals(issuer))
         .sorted()
         .findFirst();
   }
@@ -343,8 +343,12 @@ public abstract class ModerationFeatureBase extends FeatureBase implements Moder
   }
 
   // ETC.
-  private Optional<UUID> getSenderId(CommandSender sender) {
-    return Optional.ofNullable(sender instanceof Player ? ((Player) sender).getUniqueId() : null);
+  @Nullable
+  private UUID getSenderId(CommandSender sender) {
+    if (!(sender instanceof Player)) return null;
+
+    Player player = (Player) sender;
+    return player.getUniqueId();
   }
 
   private Optional<UUID> isBanEvasion(String address) {

--- a/src/main/java/dev/pgm/community/moderation/feature/types/SQLModerationFeature.java
+++ b/src/main/java/dev/pgm/community/moderation/feature/types/SQLModerationFeature.java
@@ -22,6 +22,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.logging.Logger;
 import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
 import org.bukkit.configuration.Configuration;
@@ -96,7 +97,7 @@ public class SQLModerationFeature extends ModerationFeatureBase {
   }
 
   @Override
-  public CompletableFuture<Boolean> pardon(String target, Optional<UUID> issuer) {
+  public CompletableFuture<Boolean> pardon(String target, @Nullable UUID issuer) {
     CompletableFuture<Optional<UUID>> playerId =
         NameUtils.isMinecraftName(target)
             ? getUsers().getStoredId(target)
@@ -179,7 +180,7 @@ public class SQLModerationFeature extends ModerationFeatureBase {
         if (punishment.getType() == PunishmentType.NAME_BAN) {
           String bannedName = punishment.getReason();
           if (!event.getName().equalsIgnoreCase(bannedName)) {
-            pardon(punishment.getTargetId().toString(), Optional.empty());
+            pardon(punishment.getTargetId().toString(), null);
             event.setLoginResult(Result.ALLOWED);
             logger.info(
                 String.format(
@@ -305,7 +306,7 @@ public class SQLModerationFeature extends ModerationFeatureBase {
   }
 
   @Override
-  public CompletableFuture<Boolean> unmute(UUID id, Optional<UUID> issuer) {
+  public CompletableFuture<Boolean> unmute(UUID id, @Nullable UUID issuer) {
     return service
         .unmute(id, issuer)
         .thenApplyAsync(

--- a/src/main/java/dev/pgm/community/moderation/punishments/Punishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/Punishment.java
@@ -49,14 +49,14 @@ public class Punishment implements Comparable<Punishment> {
 
   private UUID punishmentId;
   private UUID targetId;
-  private Optional<UUID> issuerId;
+  private @Nullable UUID issuerId;
   private String reason;
-  private Instant timeIssued;
+  private long timeIssued;
   private boolean active;
   private @Nullable Duration duration;
 
-  private Instant lastUpdated;
-  private Optional<UUID> lastUpdatedBy;
+  private long lastUpdated;
+  private @Nullable UUID lastUpdatedBy;
 
   private String service;
 
@@ -66,13 +66,13 @@ public class Punishment implements Comparable<Punishment> {
       PunishmentType type,
       UUID punishmentId,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
       Duration duration,
-      Instant timeIssued,
+      long timeIssued,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     this.type = type;
     this.punishmentId = punishmentId;
@@ -103,8 +103,13 @@ public class Punishment implements Comparable<Punishment> {
     return targetId;
   }
 
-  public Optional<UUID> getIssuerId() {
+  @Nullable
+  public UUID getIssuerId() {
     return issuerId;
+  }
+
+  public boolean isConsole() {
+    return getIssuerId() == null;
   }
 
   public String getReason() {
@@ -112,7 +117,7 @@ public class Punishment implements Comparable<Punishment> {
   }
 
   public Instant getTimeIssued() {
-    return timeIssued;
+    return Instant.ofEpochMilli(timeIssued);
   }
 
   public boolean isActive() {
@@ -124,10 +129,10 @@ public class Punishment implements Comparable<Punishment> {
   }
 
   public Instant getLastUpdated() {
-    return lastUpdated;
+    return Instant.ofEpochMilli(lastUpdated);
   }
 
-  public Optional<UUID> getLastUpdatedBy() {
+  public @Nullable UUID getLastUpdatedBy() {
     return lastUpdatedBy;
   }
 
@@ -169,9 +174,9 @@ public class Punishment implements Comparable<Punishment> {
           .kickPlayer(
               formatPunishmentScreen(
                   getConfig(),
-                  getIssuerId().isPresent()
-                      ? PlayerComponent.player(getIssuerId().get(), NameStyle.FANCY)
-                      : UsernameFormatUtils.CONSOLE_NAME,
+                  isConsole()
+                      ? UsernameFormatUtils.CONSOLE_NAME
+                      : PlayerComponent.player(getIssuerId(), NameStyle.FANCY),
                   silent));
       return true;
     }
@@ -187,11 +192,11 @@ public class Punishment implements Comparable<Punishment> {
     Component titleWord = translatable("misc.warning", NamedTextColor.DARK_RED);
     Component title = text().append(WARN_SYMBOL).append(titleWord).append(WARN_SYMBOL).build();
     Component subtitle;
-    if (Duration.between(timeIssued, Instant.now()).getSeconds() >= 60) {
+    if (Duration.between(getTimeIssued(), Instant.now()).getSeconds() >= 60) {
       subtitle =
           text()
               .append(
-                  TemporalComponent.relativePastApproximate(timeIssued)
+                  TemporalComponent.relativePastApproximate(getTimeIssued())
                       .color(NamedTextColor.YELLOW)
                       .append(text(": ", NamedTextColor.YELLOW)))
               .append(text(reason, NamedTextColor.GOLD))
@@ -265,10 +270,10 @@ public class Punishment implements Comparable<Punishment> {
     lines.add(
         getType()
             .getScreenComponent(
-                Duration.between(timeIssued, Instant.now()).getSeconds() >= 60
+                Duration.between(getTimeIssued(), Instant.now()).getSeconds() >= 60
                     ? text()
                         .append(
-                            TemporalComponent.relativePastApproximate(timeIssued)
+                            TemporalComponent.relativePastApproximate(getTimeIssued())
                                 .color(NamedTextColor.YELLOW)
                                 .append(text(": ", NamedTextColor.YELLOW)))
                         .append(text(reason, NamedTextColor.RED))
@@ -331,11 +336,11 @@ public class Punishment implements Comparable<Punishment> {
         punishment.getTargetId(),
         punishment.getIssuerId(),
         punishment.getReason(),
-        punishment.getTimeIssued(),
+        punishment.getTimeIssued().toEpochMilli(),
         punishment.getDuration(),
         punishment.getType(),
         punishment.isActive(),
-        punishment.getLastUpdated(),
+        punishment.getLastUpdated().toEpochMilli(),
         punishment.getLastUpdatedBy(),
         punishment.getService());
   }
@@ -343,14 +348,14 @@ public class Punishment implements Comparable<Punishment> {
   public static Punishment of(
       UUID id,
       UUID target,
-      Optional<UUID> issuer,
+      @Nullable UUID issuer,
       String reason,
-      Instant time,
+      long time,
       @Nullable Duration length,
       PunishmentType type,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     switch (type) {
       case WARN:

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/BanPunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/BanPunishment.java
@@ -2,21 +2,20 @@ package dev.pgm.community.moderation.punishments.types;
 
 import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
-import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 
 public class BanPunishment extends Punishment {
 
   public BanPunishment(
       UUID punishmentId,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.BAN,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/ExpirablePunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/ExpirablePunishment.java
@@ -4,7 +4,6 @@ import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
 
@@ -15,13 +14,13 @@ public abstract class ExpirablePunishment extends Punishment {
       PunishmentType type,
       UUID id,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       Duration length,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         type,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/KickPunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/KickPunishment.java
@@ -2,21 +2,20 @@ package dev.pgm.community.moderation.punishments.types;
 
 import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
-import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 
 public class KickPunishment extends Punishment {
 
   public KickPunishment(
       UUID id,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.KICK,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/MutePunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/MutePunishment.java
@@ -7,8 +7,8 @@ import static tc.oc.pgm.util.text.TemporalComponent.duration;
 import dev.pgm.community.moderation.punishments.PunishmentType;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
 import tc.oc.pgm.util.Audience;
@@ -18,13 +18,13 @@ public class MutePunishment extends ExpirablePunishment {
   public MutePunishment(
       UUID id,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       Duration length,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.MUTE,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/TempBanPunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/TempBanPunishment.java
@@ -2,22 +2,21 @@ package dev.pgm.community.moderation.punishments.types;
 
 import dev.pgm.community.moderation.punishments.PunishmentType;
 import java.time.Duration;
-import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 
 public class TempBanPunishment extends ExpirablePunishment {
 
   public TempBanPunishment(
       UUID id,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       Duration length,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.TEMP_BAN,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/UsernameBanPunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/UsernameBanPunishment.java
@@ -2,21 +2,20 @@ package dev.pgm.community.moderation.punishments.types;
 
 import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
-import java.time.Instant;
-import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 
 public class UsernameBanPunishment extends Punishment {
 
   public UsernameBanPunishment(
       UUID punishmentId,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.NAME_BAN,

--- a/src/main/java/dev/pgm/community/moderation/punishments/types/WarnPunishment.java
+++ b/src/main/java/dev/pgm/community/moderation/punishments/types/WarnPunishment.java
@@ -2,9 +2,9 @@ package dev.pgm.community.moderation.punishments.types;
 
 import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
-import java.time.Instant;
 import java.util.Optional;
 import java.util.UUID;
+import javax.annotation.Nullable;
 import org.bukkit.entity.Player;
 import tc.oc.pgm.util.Audience;
 
@@ -13,12 +13,12 @@ public class WarnPunishment extends Punishment {
   public WarnPunishment(
       UUID id,
       UUID targetId,
-      Optional<UUID> issuerId,
+      @Nullable UUID issuerId,
       String reason,
-      Instant timeIssued,
+      long timeIssued,
       boolean active,
-      Instant lastUpdated,
-      Optional<UUID> lastUpdatedBy,
+      long lastUpdated,
+      @Nullable UUID lastUpdatedBy,
       String service) {
     super(
         PunishmentType.WARN,

--- a/src/main/java/dev/pgm/community/moderation/services/SQLModerationService.java
+++ b/src/main/java/dev/pgm/community/moderation/services/SQLModerationService.java
@@ -17,6 +17,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nullable;
 
 public class SQLModerationService extends SQLFeatureBase<Punishment, String>
     implements ModerationQuery {
@@ -85,12 +86,10 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
                     String type = row.getString("type");
                     long time = Long.parseLong(row.getString("time"));
                     long expires = Long.parseLong(row.getString("expires"));
-                    Instant timeIssued = Instant.ofEpochMilli(time);
                     Duration length =
                         Duration.between(Instant.ofEpochMilli(time), Instant.ofEpochMilli(expires));
                     boolean active = row.get("active");
                     long lastUpdateTime = Long.parseLong(row.getString("last_updated"));
-                    Instant lastUpdate = Instant.ofEpochMilli(lastUpdateTime);
                     String lastUpdateBy = row.getString("updated_by");
                     String service = row.getString("service");
 
@@ -102,11 +101,11 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
                                 playerId,
                                 parseIssuer(issuer),
                                 reason,
-                                timeIssued,
+                                time,
                                 length,
                                 PunishmentType.valueOf(type.toUpperCase()),
                                 active,
-                                lastUpdate,
+                                lastUpdateTime,
                                 parseIssuer(lastUpdateBy),
                                 service));
                   }
@@ -143,16 +142,16 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
     return data.equalsIgnoreCase(CONSOLE_DB_NAME);
   }
 
-  private Optional<UUID> parseIssuer(String issuer) {
-    if (isConsole(issuer)) return Optional.empty();
-    return Optional.of(UUID.fromString(issuer));
+  private UUID parseIssuer(String issuer) {
+    if (isConsole(issuer)) return null;
+    return UUID.fromString(issuer);
   }
 
-  private String convertIssuer(Optional<UUID> issuer) {
-    return issuer.isPresent() ? issuer.get().toString() : CONSOLE_DB_NAME;
+  private String convertIssuer(@Nullable UUID issuer) {
+    return issuer != null ? issuer.toString() : CONSOLE_DB_NAME;
   }
 
-  public CompletableFuture<Boolean> pardon(UUID id, Optional<UUID> issuer) {
+  public CompletableFuture<Boolean> pardon(UUID id, @Nullable UUID issuer) {
     punishmentCache.invalidate(id);
     return DB.executeUpdateAsync(
             PARDON_QUERY + MULTI_PARDON_TYPE,
@@ -178,7 +177,7 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
         .thenApplyAsync(result -> result != 0);
   }
 
-  public CompletableFuture<Boolean> unmute(UUID id, Optional<UUID> issuer) {
+  public CompletableFuture<Boolean> unmute(UUID id, @Nullable UUID issuer) {
     punishmentCache.invalidate(id);
 
     return DB.executeUpdateAsync(
@@ -247,12 +246,10 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
                   String type = row.getString("type");
                   long time = Long.parseLong(row.getString("time"));
                   long expires = Long.parseLong(row.getString("expires"));
-                  Instant timeIssued = Instant.ofEpochMilli(time);
                   Duration length =
                       Duration.between(Instant.ofEpochMilli(time), Instant.ofEpochMilli(expires));
                   boolean active = row.get("active");
                   long lastUpdateTime = Long.parseLong(row.getString("last_updated"));
-                  Instant lastUpdate = Instant.ofEpochMilli(lastUpdateTime);
                   String lastUpdateBy = row.getString("updated_by");
                   String service = row.getString("service");
                   punishments.add(
@@ -261,11 +258,11 @@ public class SQLModerationService extends SQLFeatureBase<Punishment, String>
                           UUID.fromString(target),
                           parseIssuer(issuer),
                           reason,
-                          timeIssued,
+                          time,
                           length,
                           PunishmentType.valueOf(type.toUpperCase()),
                           active,
-                          lastUpdate,
+                          lastUpdateTime,
                           parseIssuer(lastUpdateBy),
                           service));
                 }

--- a/src/main/java/dev/pgm/community/network/subs/NetworkSubscriber.java
+++ b/src/main/java/dev/pgm/community/network/subs/NetworkSubscriber.java
@@ -1,6 +1,7 @@
 package dev.pgm.community.network.subs;
 
 import com.google.gson.Gson;
+import dev.pgm.community.utils.gson.GsonProvider;
 import java.util.logging.Logger;
 import redis.clients.jedis.JedisPubSub;
 
@@ -15,7 +16,7 @@ public abstract class NetworkSubscriber extends JedisPubSub {
     this.channel = channel;
     this.networkId = networkId;
     this.logger = logger;
-    this.gson = new Gson();
+    this.gson = GsonProvider.get();
   }
 
   public String getNetworkId() {

--- a/src/main/java/dev/pgm/community/network/updates/NetworkUpdateBase.java
+++ b/src/main/java/dev/pgm/community/network/updates/NetworkUpdateBase.java
@@ -1,6 +1,7 @@
 package dev.pgm.community.network.updates;
 
 import com.google.gson.Gson;
+import dev.pgm.community.utils.gson.GsonProvider;
 
 public abstract class NetworkUpdateBase<T> implements NetworkUpdate {
 
@@ -11,7 +12,7 @@ public abstract class NetworkUpdateBase<T> implements NetworkUpdate {
   public NetworkUpdateBase(T item, String channel) {
     this.item = item;
     this.channel = channel;
-    this.gson = new Gson();
+    this.gson = GsonProvider.get();
   }
 
   @Override

--- a/src/main/java/dev/pgm/community/utils/ImportUtils.java
+++ b/src/main/java/dev/pgm/community/utils/ImportUtils.java
@@ -8,6 +8,7 @@ import dev.pgm.community.moderation.ModerationConfig;
 import dev.pgm.community.moderation.punishments.Punishment;
 import dev.pgm.community.moderation.punishments.PunishmentType;
 import dev.pgm.community.users.feature.UsersFeature;
+import dev.pgm.community.utils.gson.GsonProvider;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.nio.charset.Charset;
@@ -18,14 +19,13 @@ import java.time.Instant;
 import java.util.Arrays;
 import java.util.Date;
 import java.util.List;
-import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
 import tc.oc.pgm.util.text.TextException;
 
 public class ImportUtils {
 
-  private static final Gson GSON = new Gson();
+  private static final Gson GSON = GsonProvider.get();
 
   public static List<BukkitBanEntry> getBukkitBans() throws TextException {
     File file = new File("banned-players.json");
@@ -105,14 +105,14 @@ public class ImportUtils {
       return Punishment.of(
           UUID.randomUUID(),
           getUUID(),
-          Optional.empty(),
+          null,
           getReason(),
-          getCreated(),
+          getCreated().toEpochMilli(),
           getDuration(),
           getDuration() != null ? PunishmentType.TEMP_BAN : PunishmentType.BAN,
           true,
-          getCreated(),
-          Optional.empty(),
+          getCreated().toEpochMilli(),
+          null,
           config.getService());
     }
   }

--- a/src/main/java/dev/pgm/community/utils/gson/GsonProvider.java
+++ b/src/main/java/dev/pgm/community/utils/gson/GsonProvider.java
@@ -1,0 +1,13 @@
+package dev.pgm.community.utils.gson;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import dev.pgm.community.utils.gson.types.DurationConverter;
+import java.time.Duration;
+
+public class GsonProvider {
+
+  public static Gson get() {
+    return new GsonBuilder().registerTypeAdapter(Duration.class, new DurationConverter()).create();
+  }
+}

--- a/src/main/java/dev/pgm/community/utils/gson/types/DurationConverter.java
+++ b/src/main/java/dev/pgm/community/utils/gson/types/DurationConverter.java
@@ -1,0 +1,25 @@
+package dev.pgm.community.utils.gson.types;
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import java.lang.reflect.Type;
+import java.time.Duration;
+
+public class DurationConverter implements JsonSerializer<Duration>, JsonDeserializer<Duration> {
+
+  @Override
+  public JsonElement serialize(Duration src, Type typeOfSrc, JsonSerializationContext context) {
+    return new JsonPrimitive(src.toString());
+  }
+
+  @Override
+  public Duration deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+      throws JsonParseException {
+    return Duration.parse(json.getAsString());
+  }
+}

